WordPress/Classes/ViewRelated/Views/WPRichTextGallery.swift

```diff
@@ -0,0 +1,335 @@
+
+        return (title, page)
+    }
+
+    fileprivate func setupCollectionView() {
+
+        let galleryXib = UINib(nibName: galleryCellIdentifier, bundle: Bundle.main)
+
+        collectionView.register(galleryXib, forCellWithReuseIdentifier: galleryCellIdentifier)
+        collectionView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
+
+        collectionView.backgroundColor = .clear
+        collectionView.clipsToBounds = false
+
+        collectionView.dataSource = self
+        collectionView.delegate = self
+
+        collectionView.showsHorizontalScrollIndicator = false
+    }
+    
+
+    override open func encode(with aCoder: NSCoder) {
+
+        aCoder.encode(collectionView, forKey: UICollectionView.classNameWithoutNamespaces())
+        aCoder.encode(footerStack, forKey: UIStackView.classNameWithoutNamespaces())
+
+        if let url = contentURL {
+            aCoder.encode(url, forKey: "contentURL")
+        }
+
+        if let url = linkURL {
+            aCoder.encode(url, forKey: "linkURL")
+        }
+
+        aCoder.encode(viewHeight, forKey: "viewHeight")
+
+        super.encode(with: aCoder)
+    }
+
+
+    // MARK: Public Methods
+
+    func contentSize() -> CGSize {
+        return CGSize(width: collectionView.frame.width, height: viewHeight)
+    }
+
+    func contentRatio() -> CGFloat {
+        return collectionView.frame.width / viewHeight
+    }
+}
+
+extension WPRichTextGallery: UICollectionViewDataSource {
+
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return imageAttachments.count
+    }
+
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
+        return 1
+    }
+
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: galleryCellIdentifier, for: indexPath) as! ReaderGalleryCell
+        let attachment = imageAttachments[indexPath.row]
+
+        if let url = URL(string: attachment.src) {
+            if let cachedImage = imageSource?.image(for: url, with: maxDisplaySize) {
+                cell.galleryImageView.image = cachedImage
+                attachment.maxSize = cachedImage.size
+                cell.loadingIndicator.stopAnimating()
+            } else {
+                let index = mediaArray.count
+                let indexPath = IndexPath(row: index, section: 1)
+                cell.loadingIndicator.startAnimating()
+                imageSource?.fetchImage(for: url, with: maxDisplaySize, indexPath: indexPath, isPrivate: isPrivate)
+            }
+
+            let media = GalleryMedia(cell: cell, attachment: attachment)
+            mediaArray.append(media)
+        }
+
+        return cell
+    }
+
+}
+
+extension WPRichTextGallery: UICollectionViewDelegateFlowLayout {
+    
+    func setOffsetIfNeeded() {
+        //TODO: Still Needs Fixing
```

Comment:
Yes it is, you made a nice gif of the issue I'm having with the gallery UICollectionView so a few things need to be changed in this area.  The problem I am having is that when my gallery gets it frame from the UIRichContentView/TextView it is respecting the contentInset from that textView, however this causes issues with the UICollectionView because from the design I have attempted to build it should probably be the full width of the app to prevent that disappearing issue you showed here (https://user-images.githubusercontent.com/4780/32897805-738be56a-cade-11e7-80b3-3d1aa4b5845a.gif).  So I think I have a two options, change the design or stretch the frame to be the width of the screen, use contentInsets to handle some of the padding, this should make the disappearing issue go away.  Would love some feedback/opinions on what the best options here might be from both a design and development standpoint.